app/permissions_migrations.go

```diff
@@ -0,0 +1,157 @@
+	PERMISSION_DELETE_EMOJIS                   = "delete_emojis"
+	PERMISSION_DELETE_OTHERS_EMOJIS            = "delete_others_emojis"
+	PERMISSION_MANAGE_WEBHOOKS                 = "manage_webhooks"
+	PERMISSION_MANAGE_OTHERS_WEBHOOKS          = "manage_others_webhooks"
+	PERMISSION_MANAGE_INCOMING_WEBHOOKS        = "manage_incoming_webhooks"
+	PERMISSION_MANAGE_OTHERS_INCOMING_WEBHOOKS = "manage_others_incoming_webhooks"
+	PERMISSION_MANAGE_OUTGOING_WEBHOOKS        = "manage_outgoing_webhooks"
+	PERMISSION_MANAGE_OTHERS_OUTGOING_WEBHOOKS = "manage_others_outgoing_webhooks"
+)
+
+func permissionExists(permission string) func(map[string]bool) bool {
+	return func(permissions map[string]bool) bool {
+		val, ok := permissions[permission]
+		return ok && val
+	}
+}
+
+func permissionNotExists(permission string) func(map[string]bool) bool {
+	return func(permissions map[string]bool) bool {
+		val, ok := permissions[permission]
+		return !(ok && val)
+	}
+}
+
+func permissionOr(funcs ...func(map[string]bool) bool) func(map[string]bool) bool {
+	return func(permissions map[string]bool) bool {
+		for _, f := range funcs {
+			if f(permissions) {
+				return true
+			}
+		}
+		return false
+	}
+}
+
+func permissionAnd(funcs ...func(map[string]bool) bool) func(map[string]bool) bool {
+	return func(permissions map[string]bool) bool {
+		for _, f := range funcs {
+			if !f(permissions) {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+func applyPermissionsMap(permissions []string, migrationMap permissionsMap) []string {
+	finalMap := make(map[string]bool)
+	var result []string
+	for _, permission := range permissions {
+		finalMap[permission] = true
+	}
+
+	for _, transformation := range migrationMap {
+		if transformation.On(finalMap) {
+			for _, add := range transformation.Add {
+				finalMap[add] = true
+			}
+			for _, remove := range transformation.Remove {
+				finalMap[remove] = false
+			}
+		}
+	}
+
+	for key, active := range finalMap {
+		if active {
+			result = append(result, key)
+		}
+	}
+	return result
+}
+
+func (a *App) doPermissionsMigration(key string, migrationMap permissionsMap) *model.AppError {
+	if result := <-a.Srv.Store.System().GetByName(key); result.Err == nil {
+		return nil
+	}
+
+	roles, err := a.GetRoles()
+	if err != nil {
+		return err
+	}
+
+	for _, role := range roles {
+		role.Permissions = applyPermissionsMap(role.Permissions, migrationMap)
+		if result := <-a.Srv.Store.Role().Save(role); result.Err != nil {
+			return result.Err
+		}
+	}
+
+	if result := <-a.Srv.Store.System().Save(&model.System{Name: key, Value: "true"}); result.Err != nil {
+		return result.Err
+	}
+	return nil
+}
+
+func getEmojisPermissionsSplitMigration() permissionsMap {
+	return permissionsMap{
+		permissionTransformation{
+			On:     permissionExists(PERMISSION_MANAGE_EMOJIS),
+			Add:    []string{PERMISSION_CREATE_EMOJIS, PERMISSION_DELETE_EMOJIS, PERMISSION_MANAGE_EMOJIS},
+			Remove: []string{PERMISSION_MANAGE_EMOJIS},
```

Comment:
Why adding *and* removing `PERMISSION_MANAGE_EMOJIS`?

Reply:
I think this comes from my previous code where we were maintaining our previous permission because the Canary Build idea. This is a bug.

Reply:
Fixed

Reply:
:+1: